extra/modules/greenbids-real-time-data/src/main/java/org/prebid/server/hooks/modules/greenbids/real/time/data/model/predictor/ThresholdCache.java

```diff
@@ -0,0 +1,106 @@
+package org.prebid.server.hooks.modules.greenbids.real.time.data.model.predictor;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.github.benmanes.caffeine.cache.Cache;
+import com.google.cloud.storage.Blob;
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.StorageException;
+import io.vertx.core.Future;
+import io.vertx.core.Vertx;
+import org.prebid.server.exception.PreBidException;
+import org.prebid.server.hooks.modules.greenbids.real.time.data.core.ThrottlingThresholds;
+
+import java.io.IOException;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class ThresholdCache {
+
+    String gcsBucketName;
+
+    String thresholdPath;
+
+    Cache<String, ThrottlingThresholds> cache;
+
+    Storage storage;
+
+    ObjectMapper mapper;
+
+    String thresholdsCacheKeyPrefix;
+
+    AtomicBoolean isFetching;
+
+    Vertx vertx;
+
+    public ThresholdCache(
+            String thresholdPath,
+            Storage storage,
+            String gcsBucketName,
+            ObjectMapper mapper,
+            Cache<String, ThrottlingThresholds> cache,
+            String thresholdsCacheKeyPrefix,
+            Vertx vertx) {
+        this.gcsBucketName = gcsBucketName;
+        this.thresholdPath = thresholdPath;
+        this.cache = cache;
+        this.storage = storage;
+        this.mapper = mapper;
+        this.thresholdsCacheKeyPrefix = thresholdsCacheKeyPrefix;
+        this.isFetching = new AtomicBoolean(false);
+        this.vertx = vertx;
+    }
+
+    public Future<ThrottlingThresholds> getThrottlingThresholds(String pbuid) {
+        final String cacheKey = thresholdsCacheKeyPrefix + pbuid;
+        final ThrottlingThresholds cachedThrottlingThresholds = cache.getIfPresent(cacheKey);
+
+        if (cachedThrottlingThresholds != null) {
+            return Future.succeededFuture(cachedThrottlingThresholds);
+        }
+
+        if (isFetching.compareAndSet(false, true)) {
+            try {
+                fetchAndCacheThrottlingThresholds(cacheKey);
+            } finally {
+                isFetching.set(false);
+            }
+        }
+
+        return Future.failedFuture("ThrottlingThresholds fetching in progress");
+    }
+
+    private Future<Void> fetchAndCacheThrottlingThresholds(String cacheKey) {
+        return vertx.executeBlocking(promise -> {
+            try {
+                final Blob blob = getBlob();
+                final ThrottlingThresholds throttlingThresholds = loadThrottlingThresholds(blob);
+                cache.put(cacheKey, throttlingThresholds);
+            } catch (RuntimeException e) {
+                promise.fail(e);
+            }
+        });
+
+    }
```

Comment:
```java
    private void fetchAndCacheThrottlingThresholds(String cacheKey) {
        vertx.executeBlocking(this::getBlob)
                .map(this::loadThrottlingThresholds)
                .onSuccess(thresholds -> cache.put(cacheKey, thresholds));
//                .onFailure(throwable -> logSomethings)
    }
```
P.S. Apply the same approach in the ModelRunner

Reply:
Added both `onSuccess` with cache.put and `onFailure` if fetch fails. Kept the `fetchAndCacheModelRunner` void return type. After it is done `getModelRunner` returns failedFuture with msg `fetching in progress`, which then recovered on hook call and non-changed bid request with invoction action `no_action` returned.

The same goes for threshold